### PR TITLE
rec: remove delegation only (issue #7544)

### LINKED ISSUE #7544: `delegation-only` should be removed from the recursor
- Program: Recursor
 - Issue type: Feature removal request

### Short description
As far as we know, `delegation-only` is not used on the internet. It also makes for some hairy code in the syncres. It should die.

### PR DESCRIPTION
**Pull request by xiaolong-666**
### Short description

Remove `delegation-only` from the recursor. Fixes  #7544
### Checklist

I have:
- [X] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [ ] compiled this code
- [ ] tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [ ] added or modified regression test(s)
- [ ] added or modified unit test(s)
- [ ] checked that this code was merged to master

### CODE REVIEW

**Comment by phonedph1** [comment]
I think there might be some cruft in syncres.hh and possibly some test cases too.

Stuff like: 

```
  static std::unordered_set<DNSName> s_delegationOnly;

...

  static void clearDelegationOnly()
  {
    s_delegationOnly.clear();
  }
  static void addDelegationOnly(const DNSName& name)
  {
    s_delegationOnly.insert(name);
  }
```

**Comment by phonedph1** [comment]
If you search for `s_nodelegated` you'll find a few more places where the stat is still being used.

**Comment by xiaolong-666** [comment]
> If you search for `s_nodelegated` you'll find a few more places where the stat is still being used.

Ok, may be missing part, thanks

File: pdns/pdns_recursor.cc [hunk 1]
```diff
@@ -3518,8 +3518,6 @@ static void doStats(void)
       << broadcastAccFunction<uint64_t>(pleaseGetFailedServersSize)<<", ednsmap: "
       <<broadcastAccFunction<uint64_t>(pleaseGetEDNSStatusesSize)<<endl;
     g_log<<Logger::Notice<<"stats: outpacket/query ratio "<<ratePercentage(SyncRes::s_outqueries, SyncRes::s_queries)<<"%";
-    g_log<<Logger::Notice<<", "<<ratePercentage(SyncRes::s_throttledqueries, SyncRes::s_outqueries+SyncRes::s_throttledqueries)<<"% throttled, "
```
Comment (omoerbeek) [inline pdns/pdns_recursor.cc:3521]: outpacket/query ratio and throttled reports must must stay

**Comment by omoerbeek** [comment]
"dlg-only-drops" and related stuff appears also in `rec`snmp.cc`

File: pdns/pdns_recursor.cc (hunk 1 above)
Comment (xiaolong-666) [reply pdns/pdns_recursor.cc:3521]: Ok i have modified

**Comment by omoerbeek** [comment]
The removal is incomplete, the associated `dlgOnlyDropsOID` also has to be removed

**Comment by omoerbeek** [comment]
BTW, I don't know if you are doing this, but compiling (and testing) locally on your own machine often helps.

**Comment by omoerbeek** [comment]
The SNMP regress test failure needs some investigation.

**Comment by xiaolong-666** [comment]
Sorry @omoerbeek , this is a bit difficult for me at the moment, I need a little help

**Comment by omoerbeek** [comment]
> Sorry @omoerbeek , this is a bit difficult for me at the moment, I need a little help

Help with what? Comping? Running in general or with SNMP enabled? Running with SNMP is a bit awkward to setup indeed.

**Comment by xiaolong-666** [comment]
> > Sorry @omoerbeek , this is a bit difficult for me at the moment, I need a little help
> 
> Help with what? Comping? Running in general or with SNMP enabled? Running with SNMP is a bit awkward to setup indeed.

What should I do next to complete the merge. I thought this was very simple at first, but I didn’t expect so many points to be changed.

**Comment by omoerbeek** [comment]
> > > Sorry @omoerbeek , this is a bit difficult for me at the moment, I need a little help
> > 
> > 
> > Help with what? Comping? Running in general or with SNMP enabled? Running with SNMP is a bit awkward to setup indeed.
> 
> What should I do next to complete the merge. I thought this was very simple at first, but I didn’t expect so many points to be changed.

We need to get the SNMP test working. I'm working on it, the issue is that the regerss test expects the removed stat to be avaible. I is likely better to keep the removed stat, but fix it at zero, as not so break clients expecting this value.

After that, I can do a review and merge the PR.

**Comment by omoerbeek** [comment]
Having thought about this, it is better to leave the modifications to `rec-snmp.cc` out. The code handles the absence of the actual metric correctly by reporting a zero. The MIB entry in `RECURSOR-MIB.txt`  should stay as well but the DESCRIPTION should be changed to "Obsolete".

**Comment by xiaolong-666** [comment]
> Having thought about this, it is better to leave the modifications to `rec-snmp.cc` out. The code handles the absence of the actual metric correctly by reporting a zero. The MIB entry in `RECURSOR-MIB.txt` should stay as well but the DESCRIPTION should be changed to "Obsolete".

Aha, so I should restore the change of `rec-snmp.cc`, and modify the content of "DESCRIPTION" to "Obsolete"?

**Comment by omoerbeek** [comment]
Thanks! This looks good. I'd like a squash into a single commit. Can you do that?

**Comment by xiaolong-666** [comment]
> Thanks! This looks good. I'd like a squash into a single commit. Can you do that?

Of course, so should I submit a new PR?

**Comment by omoerbeek** [comment]
> > Thanks! This looks good. I'd like a squash into a single commit. Can you do that?
> 
> Of course, so should I submit a new PR?

either that or squash locally and force-push to this PR. Do as you like best.

**Comment by xiaolong-666** [comment]
Hi @omoerbeek , i have squash locally and force-push to this PR, but ci faild...

**Comment by omoerbeek** [comment]
> Hi @omoerbeek , i have squash locally and force-push to this PR, but ci faild...

Failure is unrelated. I've restarted that test.

**Review by omoerbeek — APPROVED** [review]
Thanks for the work!

**Pull request merged**